### PR TITLE
MediaTypeHeaderValue.MatchesMediaType method added.

### LINKED ISSUE #21265: New overload of MediaTypeHeaderValue.IsSubsetOf with zero allocations
Suggested by @Tratcher in https://github.com/microsoft/reverse-proxy/pull/80#discussion_r414696103

### Is your feature request related to a problem? Please describe.
Code duplication. The gRPC server stack, and now YARP, are doing this by hand.

### Additional context
Code locations that would benefit from this:
* YARP: [GRpcProtocolHelper.IsGRpcContentType](https://github.com/microsoft/reverse-proxy/pull/80#discussion_r414696103)
* gRPC server stack: [CommonGrpcProtocolHelpers.IsContentType](https://github.com/grpc/grpc-dotnet/blob/3ce9b104524a4929f5014c13cd99ba9a1c2431d4/src/Shared/CommonGrpcProtocolHelpers.cs#L26)

### PR DESCRIPTION
**Pull request by ilkayilknur**
Summary of the changes (Less than 80 chars)
Zero allocation version of the MediaTypeHeaderValue.IsSubsetOf implementation. The method takes StringSegment to produce a result and only makes type/subtype comparison.

Closes #21265

### CODE REVIEW

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs [hunk 1]
```diff
@@ -688,6 +688,100 @@ public void TryParseStrictList_WithSomeInvalidValues_ReturnsFalse()
...
+        [InlineData("text/plain;version=v1", "Text/plain;Version=v1")]
+        [InlineData("text/plain;version=v1", "tExT/plain;version=V1")]
+        [InlineData("text/plain;version=v1", "TEXT/PLAIN;VERSION=V1")]
+        [InlineData("text/plain;charset=utf-8;foo=bar;q=0.0", "text/plain;charset=utf-8;foo=bar;q=0.0")]
+        [InlineData("text/plain;charset=utf-8;foo=bar;q=0.0", "text/plain;foo=bar;q=0.0;charset=utf-8")] // different order of parameters
+        [InlineData("text/plain;charset=utf-8;foo=bar;q=0.0", "text/*;charset=utf-8;foo=bar;q=0.0")]
+        [InlineData("text/plain;charset=utf-8;foo=bar;q=0.0", "*/*;charset=utf-8;foo=bar;q=0.0")]
+        [InlineData("application/json;v=2", "application/json;*")]
+        [InlineData("application/json;v=2;charset=utf-8", "application/json;v=2;*")]
+        public void MatchesMediaType_PositiveCases(string mediaType1, string mediaType2)
```
Comment (Tratcher) [inline src/Http/Headers/test/MediaTypeHeaderValueTest.cs:712]: How much does ordering matter here? 

Take this scenario we're trying to replace:
https://github.com/grpc/grpc-dotnet/blob/3ce9b104524a4929f5014c13cd99ba9a1c2431d4/src/Shared/CommonGrpcProtocolHelpers.cs#L26
https://github.com/grpc/grpc-dotnet/search?q=IsContentType&unscoped_q=IsContentType
Or here:
https://github.com/microsoft/reverse-proxy/blob/419c4d6502d1e8510a6dd0a0bedd67fa410ec1f2/src/ReverseProxy/Utilities/ProtocolHelper.cs#L30

You'd want one pre-parsed MediaTypeHeaderValue that you're comparing all incoming headers against. If there's a wildcard, it's always in the original value, not in the second value you're comparing to.

Comment (ilkayilknur) [reply src/Http/Headers/test/MediaTypeHeaderValueTest.cs:712]: So, MediaTypeHeaderValue.Parse("*/*").MatchesMediaType("text/plain") should return true and MediaTypeHeaderValue.Parse("text/plain").MatchesMediaType("*/*") should return false right?

**Comment by JamesNK** [comment]
Looking at this, it appears it is designed to invert the usage of MediaTypeHeaderValue.

Today you would do something like this:
```cs
MetdiaTypeHeaderValue JsonMediaType = new MediaTypeHeaderValue("application/json");

// This allocates a new MediaTypeHeaderValue
if (MediaTypeHeaderValue.TryParse(request.ContentType, out var mt))
{
    return mt.IsSubsetOf(JsonMediaType);
}

return false;
```
Now you would do something like this:
```cs
MetdiaTypeHeaderValue JsonMediaType = new MediaTypeHeaderValue("application/json");

return JsonMediaType.MatchesMediaType(request.ContentType);
```
Is that correct?

To be honest I would probably continue to use [IsContentType](https://github.com/grpc/grpc-dotnet/blob/3ce9b104524a4929f5014c13cd99ba9a1c2431d4/src/Shared/CommonGrpcProtocolHelpers.cs#L26) for performance reasons. MatchesMediaType would functionally work, but it makes many more checks as it segments the content type string, and also checks for wildcards. I don't need that functionality.

**Comment by JamesNK** [comment]
MVC has a struct version of this type called MediaType. I think that lets you avoid the need to allocate a new object. Unfortunately it lives in the Microsoft.AspNetCore.Mvc.Formatters namespace.

https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Formatters/MediaType.cs

**Comment by Tratcher** [comment]
Allocating a MetdiaTypeHeaderValue shouldn't be a concern if it's global rather than per request.

**Comment by JamesNK** [comment]
I know. My point is kind of already have a non-allocating version of this API via MediaType.

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs (hunk 1 above)
Comment (ilkayilknur) [reply src/Http/Headers/test/MediaTypeHeaderValueTest.cs:707]: Yes, I'm still interested in contributing. It looks like I need to make some changes to invert comparisons and remove unnecessary test cases. I'll let you know when I finish the implementation.

Comment (ilkayilknur) [reply src/Http/Headers/test/MediaTypeHeaderValueTest.cs:694]: Can you take a look at it again? @Tratcher

**Review by Tratcher — COMMENTED** [review]
Better, thanks. Just some cleanup items left now.

File: src/Http/Headers/src/MediaTypeHeaderValue.cs [hunk 2]
```diff
@@ -672,18 +707,80 @@ private bool MatchesSubtype(MediaTypeHeaderValue set)
...
+
+            StringSegment suffix;
+            var startOfSuffix = subType.LastIndexOf(PlusCharacter);
+            if (startOfSuffix == -1)
+            {
+                suffix = default(StringSegment);
+            }
+            else
+            {
+                suffix = subType.Subsegment(startOfSuffix + 1);
```
[the remark this answers is not in the document]
Comment (ilkayilknur) [reply src/Http/Headers/src/MediaTypeHeaderValue.cs:727]: The suffix will equal to {}. I copied that behavior from the parse method. Should I change it?

File: src/Http/Headers/src/MediaTypeHeaderValue.cs [hunk 3]
```diff
@@ -437,6 +437,33 @@ public MediaTypeHeaderValue CopyAsReadOnly()
             return other;
         }
 
+        /// <summary>
+        /// Gets a value indicating whether this <see cref="MediaTypeHeaderValue"/> is a subset of
```
[the remark this answers is not in the document]
Comment (ilkayilknur) [reply src/Http/Headers/src/MediaTypeHeaderValue.cs]: I updated the comment. Can you check?

File: src/Http/Headers/src/MediaTypeHeaderValue.cs [hunk 4]
```diff
@@ -437,6 +437,33 @@ public MediaTypeHeaderValue CopyAsReadOnly()
...
+        /// </remarks>
+        public bool MatchesMediaType(StringSegment otherMediaType)
+        {
+            if (StringSegment.IsNullOrEmpty(otherMediaType))
+            {
+                return false;
+            }
+            GetMediaTypeExpressionLength(otherMediaType, 0, out var mediaType);
+
+            // "text/plain" is a subset of "text/plain", "text/*" and "*/*". "*/*" is a subset only of "*/*".
```
[the remark this answers is not in the document]
Comment (ilkayilknur) [reply src/Http/Headers/src/MediaTypeHeaderValue.cs]: Yes. Removed the comment.

File: src/Http/Headers/src/MediaTypeHeaderValue.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Http/Headers/src/MediaTypeHeaderValue.cs:726]: Ok, I just wanted to be sure it didn't throw an index out of bounds exception.

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs [hunk 5]
```diff
@@ -683,6 +683,85 @@ public void TryParseStrictList_WithSomeInvalidValues_ReturnsFalse()
...
+
+            // Act
+            var matches = parsedMediaType1.MatchesMediaType(mediaType2);
+
+            // Assert
+            Assert.False(matches);
+        }
+
+        [Theory]
+        [InlineData("application/entity+json", "application/entity+json")]
```
Comment (Tratcher) [inline src/Http/Headers/test/MediaTypeHeaderValueTest.cs:724]: What about this?
```suggestion
        [InlineData("application/entity+json", "application/entity+json")]
        [InlineData("application/json", "application/entity+json")]
```

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs [hunk 6]
```diff
@@ -683,6 +683,85 @@ public void TryParseStrictList_WithSomeInvalidValues_ReturnsFalse()
...
+        [InlineData("application/*+json", "application/entity+json")]
+        [InlineData("application/*+json", "application/*+json")]
+        [InlineData("application/problem+json", "application/json")]
+        [InlineData("application/vnd.restful+json", "application/json")]
```
Comment (Tratcher) [inline src/Http/Headers/test/MediaTypeHeaderValueTest.cs:728]: These seem backwards. json is a superset of foo+json, not a subset of.

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs [hunk 7]
```diff
@@ -683,6 +683,85 @@ public void TryParseStrictList_WithSomeInvalidValues_ReturnsFalse()
...
+
+            // Act
+            var result = parsedMediaType1.MatchesMediaType(mediaType2);
+
+            // Assert
+            Assert.True(result);
+        }
+
+        [Theory]
+        [InlineData("application/entity+json", "application/entity+txt")]
```
Comment (Tratcher) [inline src/Http/Headers/test/MediaTypeHeaderValueTest.cs:745]:
```suggestion
        [InlineData("application/entity+json", "application/entity+txt")]
        [InlineData("application/entity+json", "application/json")]
```

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs [hunk 8]
```diff
@@ -683,6 +683,83 @@ public void TryParseStrictList_WithSomeInvalidValues_ReturnsFalse()
...
+        public void MatchesMediaTypeWithSuffixes_PositiveCases(string mediaType1, string mediaType2)
+        {
+            // Arrange
+            var parsedMediaType1 = MediaTypeHeaderValue.Parse(mediaType1);
+
+            // Act
+            var result = parsedMediaType1.MatchesMediaType(mediaType2);
+
+            // Assert
+            Assert.True(result);
```
Comment (Tratcher) [inline src/Http/Headers/test/MediaTypeHeaderValueTest.cs:741]: A secondary assertion to add here would be to create a MediaTypeHeaderValue for mediaType2 and assert that you get the same result for parsedMediaType2.IsSubsetOf(parsedMediaType1).

**Review by Tratcher — COMMENTED** [review]
I think there's some confusion around the matching direction with json and foo+json. I think for all of the tests you should get the same result for a.MatchesMediaType(b) and b.IsSubsetOf(a).

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs (hunk 6 above)
Comment (ilkayilknur) [reply src/Http/Headers/test/MediaTypeHeaderValueTest.cs:729]: test case fixes.

File: src/Http/Headers/test/MediaTypeHeaderValueTest.cs (hunk 8 above)
Comment (ilkayilknur) [reply src/Http/Headers/test/MediaTypeHeaderValueTest.cs:752]: I added IsSubsetOf result comparison to all test methods.

**Review by Tratcher — APPROVED** [review]

**Comment by Tratcher** [comment]
Thanks for sticking with it!

**Pull request merged**